src/core/mesh/qgsmeshdataset.cpp

```diff
@@ -1071,3 +1072,98 @@ void QgsMeshDatasetGroup::setIsScalar( bool isScalar )
 {
   mIsScalar = isScalar;
 }
+
+
+QgsMeshVerticesElevationDataset::QgsMeshVerticesElevationDataset( QgsMesh *mesh ):
+  mMesh( mesh )
+{}
+
+QgsMeshDatasetValue QgsMeshVerticesElevationDataset::datasetValue( int valueIndex ) const
+{
+  if ( mMesh && valueIndex < mMesh->vertexCount() )
+    return mMesh->vertex( valueIndex ).z();
+
+  return std::numeric_limits<double>::quiet_NaN();
+}
+
+QgsMeshDataBlock QgsMeshVerticesElevationDataset::datasetValues( bool isScalar, int valueIndex, int count ) const
+{
+  if ( !isScalar || !mMesh )
+    return QgsMeshDataBlock();
+
+  QgsMeshDataBlock block( QgsMeshDataBlock::ScalarDouble, count );
+  QVector<double> values( std::min( count, ( mMesh->vertexCount() - valueIndex ) ) );
+  for ( int i = 0; i < values.count(); ++i )
+    values[i] = mMesh->vertex( i ).z();
+  block.setValues( values );
+  block.setValid( true );
+  return block;
+}
+
+QgsMeshDataBlock QgsMeshVerticesElevationDataset::areFacesActive( int faceIndex, int count ) const
+{
+  QgsMeshDataBlock block( QgsMeshDataBlock::ActiveFlagInteger, std::min( count, ( mMesh->faceCount() - faceIndex ) ) );
+  block.setValid( true );
+  return block;
+}
+
+QgsMeshDatasetMetadata QgsMeshVerticesElevationDataset::metadata() const
+{
+  double min = std::numeric_limits<double>::max();
+  double max = -std::numeric_limits<double>::max();
```

Comment:
wondering how this interacts with MDAL. The creation of the elevation dataset (statistics etc) is also done in MDAL (or at least should be done for all formats) by function https://github.com/lutraconsulting/MDAL/blob/a0a9ca44197e9f082c429c1557b8eb178e415aab/mdal/mdal_utils.cpp#L608 
so you have both "Bed Elevation" and vertex."z" representing same data.

Reply:
There is no interaction with MDAL. This is a virtual dataset group no connected to MDAL. During editing, MDAL is not used anymore to provide data. Only the native mesh can have data to be exposed. This native mesh is loaded before editing from MDAL, but after editing started, MDAL is not used anymore until the end of editing.

This virtual dataset group is used to expose the vertices Z value of the native mesh during edition.